man/germinate.1
.Dd May 27, 2005
.Os Ubuntu
.ds volume-operating-system Ubuntu
.Dt GERMINATE 1
.Sh NAME
.Nm germinate
.Nd expand dependencies in a list of seed packages
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Fl S Ar source
.Op Fl s Ar dist
.Op Fl m Ar mirror
.Op Fl d Ar dist Ns \&,...
.Op Fl a Ar arch
.Op Fl c Ar component Ns \&,...
.Op Fl Fl vcs Ns = Ns Brq Li auto Ns | Ns Li bzr Ns | Ns Li git
.Op Fl Fl no\-rdepends
.Op Fl Fl no\-installer
.Sh DESCRIPTION
.Nm
is a program to help with the maintenance of large software distributions.
It takes a list of seed packages and a mirror of the distribution, and
produces outputs with the seed packages and their dependencies and
build-dependencies expanded out in full.
.Ss Seeds
The contents of the Ubuntu distribution, and others, are managed by means of
.Em seeds .
At their simplest, these are lists of packages which are considered
important to have in the main component of the distribution, without
explicitly listing all their dependencies and build-dependencies.
.Pp
Seed lists are typically divided up by category: a
.Li base
or
.Li minimal
seed might list the core set of packages required to make the system run at
all, while a
.Li desktop
seed might list the set of packages installed as part of a default desktop
installation.
.Nm
takes these seeds, adds their dependency trees, and produces an
.Em output
for each seed which contains a dependency-expanded list of package names.
These outputs may be handed on to archive maintenance or CD-building tools.
.Pp
Some seeds may
.Em inherit
from other seeds: they rely on those seeds to be installed.
For example, a
.Li desktop
seed will typically inherit from a
.Li minimal
seed.
.Nm
understands these inheritance relationships.
If a package in the
.Li desktop
seed depends on
.Sq foo ,
but
.Sq foo
is already part of the
.Li minimal
seed or dependency list, then
.Sq foo
will not be added to the
.Li desktop
output.
.Pp
Seeds are stored in text files downloaded from a given URL.
Lines not beginning with
.Sq "\ *\ "
(wiki-style list markup) are ignored.
.Pp
Seed entries may simply consist of a package name, or may include any of the
following special syntax:
.Bl -tag -width 6n
.It %
Seed entries beginning with
.Sq %
expand to all binaries from the given source package.
.It [...]
Seed entries may be followed with
.Sq " [" Ns Ar arch1 Ar arch2 ... Ns \&]
to indicate that they should only be used on the given architectures, or with
.Sq " [!" Ns Ar arch1 No ! Ns Ar arch2 ... Ns \&]
to indicate that they should not be used on the given architectures.
.It (...)
Seed entries in parentheses indicate that the seed should be treated as a
recommendation of metapackages generated from this seed, rather than as a
dependency.
.It !
Seed entries beginning with
.Sq \&!
cause the given package to be blacklisted from the given seed and any seeds
from which it inherits; this may be followed by
.Sq %
as above to blacklist all binaries from the given source package.
Note that this may result in uninstallable packages whose dependencies have
been blacklisted, so use this feature sparingly.
The purpose of a blacklist is to make it obvious when a package that is not
supposed to be installed ends up in
.Nm Ns 's
output, so that package relationships can be fixed to stop that happening.
It is not intended for the purpose of working around buggy package
relationships, and attempts to do so will not work because
.Ic apt
has no way to know about blacklist entries in seeds.
.It snap:name
Seed entries beginning with
.Sq snap:
are
.Em snap
packages.
These are different from
.Em deb
packages in that they do not have (build-)dependencies, cannot be
recommended, and do not end up in any resulting metapackages.
(If you try to recommend a snap package, it will be ignored completely.)
Snaps specified in seeds will be output in a
.Em .snaps
file named after the corresponding seed, as software processing the output of
.Nm
will typically need to treat snaps differently from debs.
.Nm
will not check remotely to see if a given snap is available, therefore seeds
are expected to explicitly list all architectures a snap is to be seeded on.
.Sq snap:
entries can also be suffixed with "/classic" to indicate that the snaps need to
be installed with classic confinement on end-user systems.
.It key: value
Some seeds also contain headers at the top of the file, in
.Dq key: value
format.
For the most part, these are not parsed by
.Nm
itself.
The Ubuntu
.Ic tasksel
package uses keys beginning with
.Sq Task\-
to define fields of similar names in its
.Pa .desc
files.
.Xr germinate\-update\-metapackage 1
uses some of these headers to reduce the need for fragile configuration;
see its documentation for further details.
.El
.Pp
A
.Pa STRUCTURE
file alongside the seeds lists their inheritance relationships.
It may also include lines beginning with
.Sq include ,
causing other collections of seeds to be included as if they were part of
the collection currently being germinated, or lines beginning with
.Sq feature ,
which set flags for the processing of seeds.
The only flag currently defined is
.Sq follow\-recommends ,
which causes
.Nm
to treat Recommends fields as if they were Depends.
(Features may also be set on a per-seed basis using lines beginning with
.Sq "\ *\ Feature:"
in the seed file; here,
.Sq no\-follow\-recommends
is also supported to allow Recommends-following to be turned off for
individual seeds.)
.Ss Build-dependencies and Sq supported
There is typically no need for a default desktop installation to contain all
the compilers and development libraries needed to build itself from source;
if nothing else, it would consume much more space.
Nevertheless, it is normally a requirement for the maintainers of a
distribution to support all the packages necessary to build that
distribution.
.Pp
.Nm
therefore does not add all the packages that result from following
build-dependencies of seed packages and of their dependencies (the
.Dq build-dependency tree )
to every output, unless they are also in the seed or in the dependency list.
Instead, it adds them to the output for the last seed in the
.Pa STRUCTURE
file, conventionally called
.Li supported .
.Pp
Like any other seed, the supported seed may contain its own list of
packages.
It is common to provide support for many software packages which are not in
the default installation, such as debugging libraries, optimised kernels,
alternative language support, and the like.
.Ss Outputs
The output files are named after the seed to which they correspond.
An additional output file is needed for supported, namely
.Sq supported+build\-depends ,
which contains the supported list and the build-depends lists of the other
seeds all joined together.
An
.Sq all
output is produced to represent the entire archive.
.Pp
Some other files are produced for occasional use by experts.
See the
.Pa README
file for full details on these.
.Sh OPTIONS
.Bl -tag -width 4n
.It Xo Fl v ,
.Fl Fl verbose
.Xc
Be more verbose when processing seeds.
.It Xo Fl S ,
.Fl Fl seed\-source Ar source Ns \&,...
.Xc
Fetch seeds from the specified sources.
The default is
.Pa http://people.canonical.com/~ubuntu-archive/seeds/ ,
or
.Pa http://bazaar.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/
if the
.Fl Fl vcs Ns = Ns bzr
option is used, or
.Pa git://git.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/+git/
if the
.Fl Fl vcs Ns = Ns git
option is used.
You may use
.Pa file://
URLs here to fetch seeds from the local file system; for example, if your
seeds are stored in
.Pa /home/username/seeds/debian.unstable ,
then you would use the options
.Fl S Ar file:///home/username/seeds/
.Fl s Ar debian.unstable .
.It Xo Fl s ,
.Fl Fl seed\-dist Ar dist
.Xc
Fetch seeds for distribution
.Ar dist .
The default is
.Li ubuntu.bionic .
.Pp
When fetching seeds from git, the part after the rightmost
.Sq \&.
character, if any, is treated as the branch name to check out; this rather
strange style is for backward compatibility.
.It Xo Fl m ,
.Fl Fl mirror Ar mirror
.Xc
Get package lists from
.Ar mirror .
The default is
.Pa http://archive.ubuntu.com/ubuntu/ .
May be supplied multiple times; the newest version of each package across
all archives will win.
.It Fl Fl source\-mirror Ar mirror
Get source package lists from
.Ar mirror .
The default is to use package lists mirrors.
May be supplied multiple times; the newest version of each source package
across all archives will win.
.It Xo Fl d ,
.Fl Fl dist Ar dist Ns \&,...
.Xc
Operate on the specified distributions.
The default is
.Li bionic .
Listing multiple distributions may be useful, for example, when examining
both a released distribution and its security updates.
.It Xo Fl a ,
.Fl Fl arch Ar arch
.Xc
Operate on architecture
.Ar arch .
The default is
.Li i386 .
.It Xo Fl c ,
.Fl Fl components Ar component Ns \&,...
.Xc
Operate on the specified components.
The default is
.Li main .
.It Fl Fl vcs Ns = Ns Brq Li auto Ns | Ns Li bzr Ns | Ns Li git
Check out seeds from a version control system rather than fetching them
directly from a URL.
Requires
.Ic bzr
or
.Ic git ,
as appropriate, to be installed.
For
.Ic bzr ,
use the branch found at
.Ar seed\-source Ns / Ns Ar seed\-dist ;
for
.Ic git ,
remove the part after the rightmost
.Sq \&.
character of
.Ar seed\-dist
and use it as the branch name to check out from
.Ar seed\-source Ns / Ns Ar remainder\-of\-seed\-dist .
For
.Ic auto ,
guess the version control system to use from
.Ar seed\-source
(trying both in ambiguous cases) and then proceed as above.
.It Fl Fl bzr
Check out seeds from the
.Ic bzr
branch found at
.Ar seed\-source Ns / Ns Ar seed\-dist
rather than fetching them directly from a URL.
Requires
.Ic bzr
to be installed.
This option is deprecated and is retained for backward compatibility; use
.Fl Fl vcs Ns = Ns bzr
instead.
.It Fl Fl no\-rdepends
Disable reverse-dependency calculations.
These calculations cause a large number of small files to be written out in
the
.Pa rdepends/
directory, and may take some time.
.It Fl Fl no\-installer
Do not consider debian-installer udeb packages.
While generally not the desired outcome, sometimes you might wish to omit consideration of installer packages when processing your seeds, perhaps if sending the output directly to the package manager on an already-installed system.
.It Fl Fl seed\-packages Ar parent Ns / Ns Ar pkg Ns \&,...
Treat each
.Ar pkg
as a seed by itself, inheriting from
.Ar parent
(i.e. assuming that all packages in the
.Ar parent
seed are already installed while calculating the additional dependencies of
.Ar pkg ) .
This allows the use of
.Nm
to calculate the dependencies of individual extra packages.
For example,
.Fl Fl seed\-packages Ar desktop Ns / Ns Ar epiphany\-browser
will create an
.Pa epiphany\-browser
output file listing the additional packages that need to be installed over
and above the
.Ar desktop
seed in order to install
.Ar epiphany\-browser .
.It Fl Fl always\-follow\-build\-depends
Always follow Build-Depends in all seeds, regardless of seed feature flags.
.El
.Sh BUGS
The wiki-style markup in seeds was inherited from an early implementation,
and is a wart.
.Pp
.Nm
can sometimes be confused by complicated situations involving the order in
which it encounters dependencies on virtual packages.
Explicit entries in seeds may be required to work around this.
.Pp
Handling of installer packages (udebs) is complicated, poorly documented,
and doesn't always work quite right: in particular, packages aren't demoted
to the supported seed when they should be.
.Sh AUTHORS
.An Scott James Remnant Aq scott@canonical.com
.An Colin Watson Aq cjwatson@canonical.com
.Pp
.An -nosplit
.Nm
is copyright \(co 2004, 2005, 2006, 2007, 2008
.An Canonical Ltd .
See the GNU General Public License version 2 or later for copying
conditions.
A copy of the GNU General Public License is available in
.Pa /usr/share/common\-licenses/GPL .


debian/changelog
germinate (2.27) unstable; urgency=medium

  [ Iain Lane ]
  * Initial support for specifying snaps in seeds.

  [ Colin Watson ]
  * Drop support for Python 3.0 and 3.1.
  * Change default distribution to bionic.

 -- Colin Watson <cjwatson@ubuntu.com>  Mon, 29 Jan 2018 18:04:40 +0000

germinate (2.26) unstable; urgency=medium

  * Add an --always-follow-build-depends option to force germinate to follow
    Build-Depends regardless of seed feature flags.
  * Change default distribution to zesty.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 26 Jan 2017 21:36:20 +0000

germinate (2.25) unstable; urgency=medium

  * Cope with malformed Built-Using fields, since we only started checking
    these recently and so older series have some errors.

 -- Colin Watson <cjwatson@ubuntu.com>  Tue, 19 Apr 2016 10:12:03 +0100

germinate (2.24) unstable; urgency=medium

  [ Dimitri John Ledkov ]
  * Create a [no-]follow-build-depends feature; follow by default.
  * Expand source package by adding the Built-Using set.  Treat all
    Built-Using as if part of the build-dependency tree.

  [ Colin Watson ]
  * Change default distribution to xenial.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 14 Apr 2016 22:45:06 +0100

germinate (2.23) unstable; urgency=medium

  * Handle Build-Depends-Arch, introduced in dpkg 1.16.4.
  * Require at least Python 2.7 and simplify some code accordingly.
  * Prefer Packages.xz and Sources.xz to other compression formats if
    possible.

 -- Colin Watson <cjwatson@ubuntu.com>  Wed, 09 Dec 2015 13:11:16 +0000

germinate (2.22) unstable; urgency=medium

  * Raise a slightly more useful exception if python-apt fails to parse a
    dependency.
  * Work around lack of build profile support in older versions of
    apt/python-apt.

 -- Colin Watson <cjwatson@debian.org>  Tue, 18 Aug 2015 16:57:42 +0100

germinate (2.21) unstable; urgency=medium

  * Fix broken germinate-update-metapackage caused by overenthusiastic
    sorting of imports.

 -- Colin Watson <cjwatson@ubuntu.com>  Tue, 19 May 2015 17:24:13 +0100

germinate (2.20) unstable; urgency=medium

  * Switch to git; adjust Vcs-* fields.
  * Add support for fetching seeds from git (LP: #1455689).
  * Change default distribution to wily.
  * Policy version 3.9.6: no changes required.

 -- Colin Watson <cjwatson@ubuntu.com>  Mon, 18 May 2015 10:07:11 +0100

germinate (2.19) unstable; urgency=medium

  * Make it possible to add a custom seed with no parents.
  * Fix TagFile.sections to use try/finally so that the temporary directory
    reliably gets cleaned up.

 -- Colin Watson <cjwatson@debian.org>  Sat, 20 Sep 2014 00:20:59 +0100

germinate (2.18) unstable; urgency=medium

  * Fix Vcs-Bzr field so that it actually works with bzr.  Adjust
    Vcs-Browser field to match.
  * Allow subclasses of SeedStructure to override a new make_seed method in
    order to slot in a different Seed implementation.

 -- Colin Watson <cjwatson@ubuntu.com>  Fri, 18 Jul 2014 16:07:01 +0100

germinate (2.17) unstable; urgency=medium

  * Bump recursion limit to 3000; apparently utopic trips over the existing
    limit (LP: #1312478).

 -- Colin Watson <cjwatson@ubuntu.com>  Fri, 25 Apr 2014 00:54:11 +0100

germinate (2.16) unstable; urgency=medium

  * Quote slashes in suite names when constructing local tag file names
    (thanks, Eugene Paskevich).
  * Change default distribution to trusty.
  * Update kubuntu-meta example in germinate-update-metapackage(1).
  * Policy version 3.9.5: no changes required.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 03 Apr 2014 15:40:02 +0100

germinate (2.15) unstable; urgency=medium

  * Use type and selector attributes of urllib.request.Request in Python >=
    3.1 rather than the deprecated get_type and get_selector accessor
    methods, fixing a build failure with Python 3.4 which removed these
    methods entirely.

 -- Colin Watson <cjwatson@ubuntu.com>  Wed, 15 Jan 2014 00:46:59 +0000

germinate (2.14) unstable; urgency=low

  * Micro-optimise lesser seeds calculation in
    Germinator._promote_dependency.
  * Amend dh_germinate_metapackage(1) to note that this program must be run
    after dh_prep.
  * Tolerate missing reasons when writing output.  This is possible when
    germinate is being run over multiple flavours for the same architecture,
    and packages are promoted differently between seeds depending on the
    flavour.  We can't generate a correct reason without a heavy performance
    cost in this case, but we can at least not crash.
  * Remove a few unnecessary "if condition: return True; else: return False"
    patterns.
  * Support multiarch (build-)dependency qualifiers (:any, :native).

 -- Colin Watson <cjwatson@ubuntu.com>  Fri, 11 Oct 2013 20:00:19 +0100

germinate (2.13) unstable; urgency=low

  * Fix dh_germinate_clean(1) and dh_germinate_metapackage(1) to recommend
    syntax compatible with debhelper 8.
  * Policy version 3.9.4:
    - Add a Vcs-Browser field.
  * Change default distribution to saucy.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 13 Jun 2013 12:50:22 +0100

germinate (2.12) unstable; urgency=low

  * Add dists option to distribution sections in
    germinate-update-metapackage configuration, making it more useful for
    released distributions.

 -- Colin Watson <cjwatson@ubuntu.com>  Wed, 28 Nov 2012 16:13:58 +0000

germinate (2.11) unstable; urgency=low

  * Use collections.defaultdict instead of manually initialising elements.
  * Simplify some sorted() calls, relying on the default iterator for
    mappings being .keys().
  * Simplify code to clean up bzr cache, removing an unnecessary helper
    function.
  * Add support for reading seeds from relative filesystem paths (thanks,
    Sjoerd Simons; LP: #1010186).
  * Only allow dh_python2 to operate on python-germinate, to avoid chaos
    caused by shebang rewriting.

 -- Colin Watson <cjwatson@ubuntu.com>  Tue, 18 Sep 2012 10:12:41 +0100

germinate (2.10) unstable; urgency=low


germinate.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
"""Update list files from the Wiki."""

# Copyright (c) 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011
#               Canonical Ltd.
#
# Germinate is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any
# later version.
#
# Germinate is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Germinate; see the file COPYING.  If not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301, USA.

import os
import shutil
import sys
import optparse
import logging

from Germinate import Germinator
import Germinate.Archive
import Germinate.defaults
from Germinate.seeds import Seed, SeedError, SeedStructure
import Germinate.version


def parse_options():
    parser = optparse.OptionParser(prog='germinate',
                                   version=Germinate.version.VERSION)
    parser.add_option('-v', '--verbose', dest='verbose', action='store_true',
                      default=False,
                      help='be more verbose when processing seeds')
    parser.add_option('-S', '--seed-source', dest='seeds', metavar='SOURCE',
                      help='fetch seeds from SOURCE (default: %s)' %
                           Germinate.defaults.seeds)
    parser.add_option('-s', '--seed-dist', dest='release', metavar='DIST',
                      default=Germinate.defaults.release,
                      help='fetch seeds for distribution DIST '
                           '(default: %default)')
    parser.add_option('-m', '--mirror', dest='mirrors', action='append',
                      metavar='MIRROR',
                      help='get package lists from MIRROR (default: %s)' %
                           Germinate.defaults.mirror)
    parser.add_option('--source-mirror', dest='source_mirrors',
                      action='append', metavar='MIRROR',
                      help='get source package lists from mirror '
                           '(default: value of --mirror)')
    parser.add_option('-d', '--dist', dest='dist',
                      default=Germinate.defaults.dist,
                      help='operate on distribution DIST (default: %default)')
    parser.add_option('-a', '--arch', dest='arch',
                      default=Germinate.defaults.arch,
                      help='operate on architecture ARCH (default: %default)')
    parser.add_option('-c', '--components', dest='components',
                      default='main,restricted', metavar='COMPS',
                      help='operate on components COMPS (default: %default)')
    parser.add_option('--bzr', dest='bzr', action='store_true', default=False,
                      help='fetch seeds using bzr (requires bzr to be '
                           'installed)')
    parser.add_option('--cleanup', dest='cleanup', action='store_true',
                      default=False,
                      help="don't cache Packages or Sources files")
    parser.add_option('--no-rdepends', dest='want_rdepends',
                      action='store_false', default=True,
                      help='disable reverse-dependency calculations')
    parser.add_option('--no-installer', dest='installer', action='store_false',
                      default=True,
                      help='do not consider debian-installer udeb packages')
    parser.add_option('--seed-packages', dest='seed_packages',
                      metavar='PARENT/PKG,PARENT/PKG,...',
                      help='treat each PKG as a seed by itself, inheriting '
                           'from PARENT')
    options, _ = parser.parse_args()

    if options.seeds is None:
        if options.bzr:
            options.seeds = Germinate.defaults.seeds_bzr
        else:
            options.seeds = Germinate.defaults.seeds
    options.seeds = options.seeds.split(',')

    if options.mirrors is None:
        options.mirrors = [Germinate.defaults.mirror]

    def canonicalise_mirror(mirror):
        if not mirror.endswith('/'):
            mirror += '/'
        return mirror

    options.mirrors = map(canonicalise_mirror, options.mirrors)
    if options.source_mirrors is not None:
        options.source_mirrors = map(canonicalise_mirror,
                                     options.source_mirrors)

    options.dist = options.dist.split(',')
    options.components = options.components.split(',')
    if options.seed_packages is None:
        options.seed_packages = []
    else:
        options.seed_packages = options.seed_packages.split(',')

    return options


def main():
    options = parse_options()

    logger = logging.getLogger()
    if options.verbose:
        logger.setLevel(logging.DEBUG)
    else:
        logger.setLevel(Germinator.PROGRESS)
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(logging.Formatter('%(levelname)s%(message)s'))
    logger.addHandler(handler)

    g = Germinator(options.arch)

    archive = Germinate.Archive.TagFile(
        options.dist, options.components, options.arch,
        options.mirrors, source_mirrors=options.source_mirrors,
        installer_packages=options.installer, cleanup=options.cleanup)
    g.parseSections(archive)

    if os.path.isfile("hints"):
        with open("hints") as hints:
            g.parseHints(hints)

    try:
        with Seed(options.seeds, options.release, "blacklist",
                  options.bzr) as blacklist:
            g.parseBlacklist(blacklist)
    except SeedError:
        pass

    try:
        structure = SeedStructure(options.release, options.seeds, options.bzr)
        for seed_package in options.seed_packages:
            parent, pkg = seed_package.split('/')
            structure.add(pkg, [" * " + pkg], parent)
        g.plantSeeds(structure)
    except SeedError:
        sys.exit(1)

    g.prune()
    g.grow()
    g.addExtras(structure)
    if options.want_rdepends:
        g.reverseDepends()

    for seedname in structure.names:
        g.writeList(seedname, seedname,
                    set(g.seed[seedname]) | set(g.seedrecommends[seedname]) |
                    set(g.depends[seedname]))
        g.writeList(seedname, seedname + ".seed",
                    g.seed[seedname])
        g.writeList(seedname, seedname + ".seed-recommends",
                    g.seedrecommends[seedname])
        g.writeList(seedname, seedname + ".depends",
                    g.depends[seedname])
        g.writeList(seedname, seedname + ".build-depends",
                    g.build_depends[seedname])

        if seedname != "extra" and seedname in structure.texts:
            g.writeSeedText(seedname + ".seedtext", structure.texts[seedname])
            g.writeSourceList(seedname + ".sources",
                              g.sourcepkgs[seedname])
        g.writeSourceList(seedname + ".build-sources",
                          g.build_sourcepkgs[seedname])

    all_bins = set()
    sup_bins = set()
    all_srcs = set()
    sup_srcs = set()
    for seedname in structure.names:
        if seedname == "extra":
            continue

        all_bins.update(g.seed[seedname])
        all_bins.update(g.seedrecommends[seedname])
        all_bins.update(g.depends[seedname])
        all_bins.update(g.build_depends[seedname])
        all_srcs.update(g.sourcepkgs[seedname])
        all_srcs.update(g.build_sourcepkgs[seedname])

        if seedname == g.supported:
            sup_bins.update(g.seed[seedname])
            sup_bins.update(g.seedrecommends[seedname])
            sup_bins.update(g.depends[seedname])
            sup_srcs.update(g.sourcepkgs[seedname])

        # Only include those build-dependencies that aren't already in the
        # dependency outputs for inner seeds of supported. This allows
        # supported+build-depends to be usable as an "everything else"
        # output.
        build_depends = dict.fromkeys(g.build_depends[seedname], True)
        build_sourcepkgs = dict.fromkeys(g.build_sourcepkgs[seedname], True)
        for seed in g.innerSeeds(g.supported):
            build_depends.update(dict.fromkeys(g.seed[seed], False))
            build_depends.update(dict.fromkeys(g.seedrecommends[seed], False))
            build_depends.update(dict.fromkeys(g.depends[seed], False))
            build_sourcepkgs.update(dict.fromkeys(g.sourcepkgs[seed], False))
        sup_bins.update([k for (k, v) in build_depends.iteritems() if v])
        sup_srcs.update([k for (k, v) in build_sourcepkgs.iteritems() if v])

    g.writeList("all", "all", all_bins)
    g.writeSourceList("all.sources", all_srcs)

    g.writeList("all", "%s+build-depends" % g.supported, sup_bins)
    g.writeSourceList("%s+build-depends.sources" % g.supported, sup_srcs)

    g.writeList("all", "all+extra", g.all)
    g.writeSourceList("all+extra.sources", g.all_srcs)

    g.writeProvidesList("provides")

    structure.write("structure")
    structure.writeDot("structure.dot")

    if os.path.exists("rdepends"):
        shutil.rmtree("rdepends")
    if options.want_rdepends:
        os.mkdir("rdepends")
        os.mkdir(os.path.join("rdepends", "ALL"))
        for pkg in g.all:
            dirname = os.path.join("rdepends", g.packages[pkg]["Source"])
            if not os.path.exists(dirname):
                os.mkdir(dirname)

            g.writeRdependList(os.path.join(dirname, pkg), pkg)
            os.symlink(os.path.join("..", g.packages[pkg]["Source"], pkg),
                       os.path.join("rdepends", "ALL", pkg))

    g.writeBlacklisted("blacklisted")

if __name__ == "__main__":
    main()
